### PR TITLE
overlord/snapstate: start using revisions higher than 100000 for local installs (sideloads)

### PR DESCRIPTION
**Pull request by pedronis**
This also fixes InstallPath to check whether there was a previous snap of the same name and do the proper unlinking etc...

### CODE REVIEW

File: overlord/snapstate/snapmgr.go [hunk 1]
```diff
@@ -58,11 +58,12 @@ func (ss *SnapSetup) MountDir() string {
...
-	Candidate *snap.SideInfo   `josn:"candidate,omitempty"`
-	Active    bool             `json:"active,omitempty"`
-	Channel   string           `json:"channel,omitempty"`
-	DevMode   bool             `json:"dev-mode,omitempty"`
+	Sequence             []*snap.SideInfo `json:"sequence"` // Last is current
+	Candidate            *snap.SideInfo   `josn:"candidate,omitempty"`
+	Active               bool             `json:"active,omitempty"`
+	Channel              string           `json:"channel,omitempty"`
+	DevMode              bool             `json:"dev-mode,omitempty"`
+	LastSideloadRevision int              `json:"last-sideload-revision,omitempty"`
```
Comment (niemeyer) [inline overlord/snapstate/snapmgr.go]: Can we please call this "generated-revision" or similar? The term "sideload" still feels somewhat awkward to me in this context, and I suspect the introduction of assertions will blur those lines further and make it even less friendly. Over time I think the term "devmode" will take over.

Comment (niemeyer) [inline overlord/snapstate/snapmgr.go]: (shortened suggestion to "gen-revision")

Comment (niemeyer) [inline overlord/snapstate/snapmgr.go]: (changed suggestion again to "local-revision")

File: overlord/snapstate/snapmgr.go [hunk 2]
```diff
@@ -124,8 +127,24 @@ func (m *SnapManager) doPrepareSnap(t *state.Task, _ *tomb.Tomb) error {
...
 	}
 
+	if ss.Revision == 0 { // sideloading
+		// to not clash with not sideload installs
+		// and to not have clashes between them
+		// use incremental revisions starting at 100000
+		// for sideloads
+		revision := snapst.LastSideloadRevision
+		if revision == 0 {
+			revision = firstSideloadRevision
```
Comment (niemeyer) [inline overlord/snapstate/snapmgr.go]: Can we please +1 the base, so we start those on *1 too, same as the store revisions.

File: overlord/snapstate/snapstate.go [hunk 3]
```diff
@@ -146,7 +135,19 @@ func Install(s *state.State, name, channel string, flags snappy.InstallFlags) (*
...
 func InstallPath(s *state.State, path, channel string, flags snappy.InstallFlags) (*state.TaskSet, error) {
-	return doInstall(s, false, "", path, channel, flags)
+	info, err := readSnapInfo(path)
+	if err != nil {
+		return nil, err
+	}
+	snapName := info.Name()
+
+	var snapst SnapState
+	err = Get(s, snapName, &snapst)
```
Comment (niemeyer) [inline overlord/snapstate/snapstate.go]: What do we need this for?

Comment (pedronis) [inline overlord/snapstate/snapstate.go]: to know snapst.Active (for the name) to know if we need to unlink

Comment (pedronis) [inline overlord/snapstate/snapstate.go]: the flag goes into doInstall that adds a unlink-snap task for that case

Comment (pedronis) [inline overlord/snapstate/snapstate.go]: that's what the 2nd test I added is all about

File: overlord/snapstate/snapmgr.go (hunk 2 above)
Comment (pedronis) [inline overlord/snapstate/snapmgr.go]: so first sideload revision to be 100001 ?

File: overlord/snapstate/snapstate.go (hunk 3 above)
Comment (niemeyer) [inline overlord/snapstate/snapstate.go]: Sorry, I'm blind.

**Comment by niemeyer** [comment]
LGTM

**Pull request merged**